src/main/java/bio/terra/service/dataset/flight/create/DeleteDatasetAssetStep.java

```diff
@@ -17,17 +20,17 @@ public DeleteDatasetAssetStep(AssetDao assetDao) {
     this.assetDao = assetDao;
   }
 
-  private UUID getAssetId(FlightContext context) {
-    // get asset Id
-    UUID assetId =
-        UUID.fromString(
-            context.getInputParameters().get(JobMapKeys.ASSET_ID.getKeyName(), String.class));
-    return assetId;
-  }
-
   @Override
   public StepResult doStep(FlightContext context) {
-    assetDao.delete(getAssetId(context));
+    FlightMap inputs = context.getInputParameters();
+    UUID datasetId = UUID.fromString(inputs.get(JobMapKeys.DATASET_ID.getKeyName(), String.class));
+    UUID assetId = UUID.fromString(inputs.get(JobMapKeys.ASSET_ID.getKeyName(), String.class));
+    List<UUID> snapshotIds = assetDao.retrieveSnapshotsForAsset(datasetId, assetId);
+    if (snapshotIds.size() != 0) {
+      throw new ValidationException(
+          "The asset is being used by snapshots: " + StringUtils.join(snapshotIds, ","));
+    }
+    assetDao.delete(assetId);
```

Comment:
Is there a race condition here, If another flight is creating a snapshot while this flight is running? Is there a way to both check for the constraint and delete the asset in a single transaction, so any changes to the affected tables would cause the transaction to restart?

Reply:
That's a good point. I don't DB transactionality will necessarily help here.  I think we may want to put an exclusive lock on the dataset during the flight that calls this and that should protect us